tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/internal/findandreplace/FindReplaceLogicTest.java

```diff
@@ -547,6 +551,32 @@ public void testWholeWordSearchAvailable() {
 		assertThat(findReplaceLogic.isWholeWordSearchAvailable(""), is(false));
 	}
 
+
+	@Test
+	public void testCanReplaceAfterWrap() {
+		TextViewer textViewer= setupTextViewer(LINE_STRING + lineSeparator() + LINE_STRING);
+		textViewer.setSelectedRange(LINE_STRING_LENGTH + lineSeparator().length(), 0);
+		IFindReplaceLogic findReplaceLogic= setupFindReplaceLogicObject(textViewer);
+		findReplaceLogic.activate(SearchOptions.FORWARD);
+		findReplaceLogic.activate(SearchOptions.WRAP);
+		findReplaceLogic.performSelectAndReplace(LINE_STRING, "");
+		assertThat(textViewer.getTextWidget().getText(), is(LINE_STRING + lineSeparator()));
+		findReplaceLogic.performSelectAndReplace(LINE_STRING, "");
+		assertThat(textViewer.getTextWidget().getText(), is(lineSeparator()));
+	}
+
+	@Test
+	public void onlySelectAndReplacesIfFindSuccessful() {
+		String setupString= "ABCD" + lineSeparator() + LINE_STRING;
```

Comment:
Should't this string contain the substring "NOTFOUND"? Otherwise the replace operation cannot replace anything anyway, no matter whether a specific range is selected or not.

Reply:
If this string contained the subtring "NOTFOUND", then the substring "NOTFOUND" would be found, meaning that there is a selection that can be set for which the replacement would be valid.

The point of this test is that we want the replace operation to recognize itself when it cannot replace anything (e.g. because the search-string could not be found) and then leaves the current selection unaffected. In case the find/replace-operation cannot replace anything by itself, we don't want it to call the internal "replace selection"-method since that would (... if implemented as specified ...) overwrite a selection which doesn't contain the search-string

Reply:
Thank you for the clarification. I accidentally though the selection would introduce a search scoped, but actually the "global" option is not deactivated.

Still I do not understand how this test is supposed to validate that the selection did not change. Currently, it only validates that when performing a search and replace operation on a string that does not contain the string to replace, then the content does not change. But I do not find any assertion that validated that the selection does not change.

Reply:
> this test tests that the text highlighted by the selection does not change.

I am somehow stuck here. Can you elaborate how the test ensures this? The only assertion of this test validates that the text of the target does not change, but that does not validate anything with respect to the selection/highlighting, does it?

Reply:
> I believe the misunderstanding stems from me not checking for a selection being set after the replace.

Exactly, because I do not see why it should be necessary to set a selection in the test setup when this has no effect to the rest of the test (i.e., affecting the test outcome). Thanks for adding the selection checks.